subsys/fs/ext2/ext2_impl.c

```diff
@@ -0,0 +1,1542 @@
+		bool last_entry = next_path[0] == '\0';
+
+		if (!last_entry) {
+			/*  prepare the next loop iteration */
+
+			if (ino < 0) {
+				/* next entry not found */
+				ret = -ENOENT;
+				goto out;
+			}
+
+			rc = ext2_inode_get(fs, ino, &next);
+			if (rc < 0) {
+				/* error while fetching next entry */
+				ret = rc;
+				goto out;
+			}
+
+			if (!(next->i_mode & EXT2_S_IFDIR)) {
+				/* path component should be directory */
+				ret = -ENOTDIR;
+				goto out;
+			}
+
+			/* Go to the next path component */
+			path = next_path;
+
+			/* Move to next directory */
+			ext2_inode_drop(cur_dir);
+			cur_dir = next;
+
+			next = NULL;
+			continue;
+		}
+
+		/* Last entry */
+
+		if (ino < 0 && !(args->flags & LOOKUP_ARG_CREATE)) {
+			/* entry not found but we need it */
+			ret = -ENOENT;
+			goto out;
+		}
+
+		if (ino > 0) {
+			rc = ext2_inode_get(fs, ino, &next);
+			if (rc < 0) {
+				ret = rc;
+				goto out;
+			}
+		}
+
+		/* Store parent directory and offset in parent directory */
+		if (args->flags & (LOOKUP_ARG_CREATE | LOOKUP_ARG_STAT | LOOKUP_ARG_UNLINK)) {
+			/* In create it will be valid only if we have found existing file */
+			args->offset = dir_off;
+			args->parent = cur_dir;
+			cur_dir = NULL;
+		}
+
+		/* Store name info */
+		if (args->flags & LOOKUP_ARG_CREATE) {
+			args->name_pos = path - args->path;
+			args->name_len = len;
+		}
+
+		/* Store found inode */
+		if (ino > 0) {
+			args->inode = next;
+			next = NULL;
+		}
+
+		ret = 0;
+		goto out;
+	}
+
+out:
+	/* Always free that inodes.
+	 * If some of them is returned from function then proper pointer is set to NULL.
+	 */
+	ext2_inode_drop(cur_dir);
+	ext2_inode_drop(next);
+	return ret;
+}
+
+static const char *skip_slash(const char *str)
+{
+	while (*str != '\0') {
+		if (*str != '/') {
+			return str;
+		}
+		str++;
+	}
+	return str;
+}
+
+/* Return position of given char or end of string. */
+static char *strchrnul(const char *str, const char c)
+{
+	char *p = strchr(str, c);
+
+	if (!p) {
+		p = strchr(str, '\0');
+	}
+	return p;
+}
+
+/**
+ * @brief Find inode
+ *
+ * @note Inodes are 32 bit. When we return signed 64 bit number then we don't
+ *       lose any information.
+ *
+ * @param r_offset If not NULL then offset in directory of that entry is written here.
+ * @return Inode number or negative error code
+ */
+static int64_t find_dir_entry(struct ext2_inode *inode, const char *name, size_t len,
+		uint32_t *r_offset)
+{
+	int rc;
+	uint32_t block, block_off, offset = 0;
+	struct ext2_data *fs = inode->i_fs;
+
+	while (offset < inode->i_size) {
+		block = offset / fs->block_size;
+		block_off = offset % fs->block_size;
+
+		rc = ext2_fetch_inode_block(inode, block);
+		if (rc < 0) {
+			return rc;
+		}
+
+		struct ext2_disk_dentry *de =
+			(struct ext2_disk_dentry *)(inode_current_block_mem(inode) + block_off);
+		size_t maxlen = MAX(len, (size_t)de->de_name_len);
+
+		if (strncmp(de->de_name, name, maxlen) == 0) {
```

Comment:
Is this correct? If `len` does not match, there is no point for comparison, it is not the same name.